reps/2023-03-10-batch-inference.md

```diff
@@ -120,6 +119,110 @@ predictions = ds.read_parquet("s3://data")
     .map_predictor(predictor)
 ```
 
+## Example: Implementing a custom Predictor.
+Using Checkpoints are no longer necessary.
+
+### Before
+```python
+class CustomPredictor(Predictor):
+    def __init__(
+      self,
+      net: gluon.Block,
+      preprocessor: Optional[Preprocessor] = None,
+    ):
+        self.net = net
+        super().__init__(preprocessor)
+
+    @classmethod
+    def from_checkpoint(
+        cls,
+        checkpoint: Checkpoint,
+        net: gluon.Block,
+    ) -> Predictor:
+        with checkpoint.as_directory() as directory:
+            path = os.path.join(directory, "net.params")
+            net.load_parameters(path)
+        return cls(net, preprocessor=checkpoint.get_preprocessor())
+
+    def _predict_numpy(
+        self,
+        data: Union[np.ndarray, Dict[str, np.ndarray]],
+        dtype: Optional[np.dtype] = None,
+    ) -> Dict[str, np.ndarray]:
+        # If `data` looks like `{"features": array([...])}`, unwrap the `dict` and pass
+        # the array directly to the model.
+        if isinstance(data, dict) and len(data) == 1:
+            data = next(iter(data.values()))
+
+        inputs = mx.nd.array(data, dtype=dtype)
+        outputs = self.net(inputs).asnumpy()
+
+        return {"predictions": outputs}
+
+dataset = ray.data.read_images(
+    "s3://anonymous@air-example-data-2/imagenet-sample-images", size=(224, 224)
+)
+
+
+def preprocess(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+    # (B, H, W, C) -> (B, C, H, W)
+    batch["image"] = batch["image"].transpose(0, 3, 1, 2)
+    return batch
+
+
+# Create the preprocessor and set it in the checkpoint.
+# This preprocessor will be used to transform the data prior to prediction.
+preprocessor = BatchMapper(preprocess, batch_format="numpy")
+checkpoint.set_preprocessor(preprocessor=preprocessor)
+
+
+net = gluon.model_zoo.vision.resnet50_v1(pretrained=True)
+predictor = BatchPredictor.from_checkpoint(
+    checkpoint, MXNetPredictor, net=net
+)
+predictor.predict(dataset)
+```
+
+### After
+```python
+class MXNetPredictor(Predictor):
+    def __init__(
+      self,
+      net: gluon.Block,
+    ):
+        self.net = net
+        super().__init__()
+
+    def _predict_numpy(
```

Comment:
Hmm this is a bit confusing, since you specify "numpy" both here and in the batch format arg. If one is mismatched, I guess you get an error?

Maybe it would be better to have just recommending overriding the predict() method.

Reply:
oh no, the `batch_format` arg is specified in `map_batches` for the preprocessing logic. `map_predictor` doesn't take in a batch format.

Reply:
I guess I'm saying--- in this example (and other examples we create), why not override the plain `predict` class instead? That way, the term "numpy" shows up only once, and the user doesn't need to understand the automatic batch format choosing magic in the Predictor class.

Reply:
Not sure if I'm understanding...we still want to provide both a numpy path and a pandas path right?

When implementing a custom predictor, the user can pick one (or both) to implement. The term numpy is still showing up only in the Predictor. `map_predictor` does not take in a `batch_format` argument.

Reply:
Stepping back, I'm kind of saying the entire way we have designed this is wrong...

Predictor shouldn't query dataset for it's "dataset format". The predictor should say, "this is the batch format I need", and the dataset should internally optimize to minimize format conversions.

In this example, the user ideally should be able to simply write:

```
class MyPrep(Preprocessor):
   batch_format = "numpy"
   
   def predict(self, data):
        assert data is in numpy format
```

Reply:
In other words: why is this so much more complicated than the map_batches() signature? It shouldn't be.

Reply:
Yes completely agree on what you are saying. But I don't think the API needs to change to achieve this.

> The predictor should say, "this is the batch format I need"

This is achieved by the Predictor either implementing `_predict_pandas` or `_predict_numpy`.

I think the question is should this distinction be in the form of 2 separate methods, or overloading onto a single `predict` method, and having a `batch_format` class attribute.

Personally, I slightly prefer the former just to be more explicit.

Reply:
Either way, we will be removing the dataset format querying. But this is more of an implementation change than an API change.

Reply:
OK, it seems fine for now if we remove batch format from map_predictor().